File: tensorflow_serving/example/xgboost_client.py
# Copyright 2016 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

#!/usr/bin/env python2.7

"""A client that talks to tensorflow_model_server.

Typical usage example:

    xgboost_client.py --num_tests=100 --server=localhost:9000
"""

from __future__ import print_function

import sys
import time
import grpc
import numpy
import random
from grpc._cython.cygrpc import CompressionAlgorithm
from grpc._cython.cygrpc import CompressionLevel

from tensorflow_serving.apis import predict_pb2
from tensorflow_serving.apis import prediction_service_pb2
from tensorflow_serving.apis import prediction_service_pb2_grpc
from google.protobuf import text_format

import tensorflow as tf

tf.compat.v1.app.flags.DEFINE_integer('num_tests', 1, 'Number of tests')
tf.compat.v1.app.flags.DEFINE_string('server', '', 'PredictionService host:port')
FLAGS = tf.compat.v1.app.flags.FLAGS


def do_inference(hostport, num_tests):
  """Tests PredictionService with requests.

  Args:
    hostport: Host:port address of the PredictionService.
    num_tests: Number of test images to use.

  Returns:
    void.
  """
  host, port = hostport.split(':')
  options = [
     ("grpc.default_compression_algorithm", CompressionAlgorithm.gzip),
     ("grpc.grpc.default_compression_level", CompressionLevel.high)
  ]
  channel = grpc.insecure_channel(hostport, options)
  stub = prediction_service_pb2_grpc.PredictionServiceStub(channel)
  for _ in range(num_tests):
    request = predict_pb2.PredictRequest()
    request.model_spec.name = 'test'
    # request.model_spec.version.value = 1
    xgboost_feature_score_1 = predict_pb2.FeatureScore(
      id=[2, 34, 2000, 2206],
      score=[1, 1, 0.646667, 0.727273],
    )
    xgboost_features =  [xgboost_feature_score_1 for i in range(0, 1)]
    xgboost_feature_vector = predict_pb2.FeatureScoreVector(feature_score=xgboost_features)
    request.inputs['xgboost_features'].CopyFrom(xgboost_feature_vector)
    response = stub.Predict(request, 30.0)
    print(response)

def main(_):
  if FLAGS.num_tests > 10000:
    print('num_tests should not be greater than 10k')
    return
  if not FLAGS.server:
    print('please specify server host:port')
    return
  do_inference(FLAGS.server, FLAGS.num_tests)


if __name__ == '__main__':
  tf.compat.v1.app.run()

File: tensorflow_serving/example/xgboost_client.cc
/* Copyright 2017 Google Inc. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include <iostream>

#include "grpcpp/create_channel.h"
#include "grpcpp/security/credentials.h"
#include "google/protobuf/map.h"
#include "tensorflow/core/framework/tensor.h"
#include "tensorflow/core/platform/types.h"
#include "tensorflow/core/util/command_line_flags.h"
#include "tensorflow_serving/apis/prediction_service.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;

using tensorflow::serving::PredictRequest;
using tensorflow::serving::PredictResponse;
using tensorflow::serving::PredictionService;
using tensorflow::serving::FeatureScore;
using tensorflow::serving::FeatureScoreVector;

typedef google::protobuf::Map<tensorflow::string, tensorflow::TensorProto> OutMap;

class ServingClient {
 public:
  ServingClient(std::shared_ptr<Channel> channel)
      : stub_(PredictionService::NewStub(channel)) {}

  tensorflow::string callPredict(const tensorflow::string& model_name) {
    PredictRequest predictRequest;
    PredictResponse response;
    ClientContext context;

    predictRequest.mutable_model_spec()->set_name(model_name);

    google::protobuf::Map<tensorflow::string, FeatureScoreVector>& inputs =
        *predictRequest.mutable_inputs();

    // xgboost features
    FeatureScoreVector xgboost_feature_score_vector;
    auto xgboost_feature_score = xgboost_feature_score_vector.add_feature_score();
    xgboost_feature_score->add_id(2);
    xgboost_feature_score->add_score(1);
    xgboost_feature_score->add_id(34);
    xgboost_feature_score->add_score(1);
    xgboost_feature_score->add_id(2000);
    xgboost_feature_score->add_score(0.64667);
    xgboost_feature_score->add_id(2206);
    xgboost_feature_score->add_score(0.727273);
    inputs["xgboost_features"] = xgboost_feature_score_vector;
    
    Status status;
    status = stub_->Predict(&context, predictRequest, &response);

    if (status.ok()) {
      std::cout << "call predict ok" << std::endl;
      std::cout << "outputs size is " << response.outputs_size() << std::endl;
      OutMap& map_outputs = *response.mutable_outputs();
      if(map_outputs.contains("index")) {
        tensorflow::TensorProto& result_tensor_proto = map_outputs.at("index");
        tensorflow::Tensor tensor;
        bool converted = tensor.FromProto(result_tensor_proto);
        if (converted) {
          std::cout << "the result tensor[" << 0
                    << "] is:" << std::endl
                    << tensor.SummarizeValue(tensor.dim_size(0)) << std::endl;
        } else {
          std::cout << "the result tensor[" << 0
                    << "] convert failed." << std::endl;
        }
      }
      return "Done.";
    } else {
      std::cout << "gRPC call return code: " << status.error_code() << ": "
                << status.error_message() << std::endl;
      return "gRPC failed.";
    }
  }

 private:
  std::unique_ptr<PredictionService::Stub> stub_;
};

int main(int argc, char** argv) {
  tensorflow::string server_port = "localhost:8500";
  tensorflow::string model_name = "test";
  std::vector<tensorflow::Flag> flag_list = {
      tensorflow::Flag("server_port", &server_port,
                       "the IP and port of the server"),
      tensorflow::Flag("model_name", &model_name, "name of model")};
  tensorflow::string usage = tensorflow::Flags::Usage(argv[0], flag_list);
  const bool parse_result = tensorflow::Flags::Parse(&argc, argv, flag_list);
  ServingClient guide(grpc::CreateChannel(server_port, grpc::InsecureChannelCredentials()));
  std::cout<<guide.callPredict(model_name)<<std::endl;
  return 0;
}
